relay-client/src/client.rs

```diff
@@ -0,0 +1,749 @@
+        dst_id: String,
+        mode: Mode,
+    ) -> Self {
+        Self {
+            message_buffer: Arc::new(Mutex::new(VecDeque::new())),
+            outgoing_tx: None,
+            command_tx: None,
+            shutdown_token: None,
+            shutdown_completed: None,
+            config: Config {
+                src_id,
+                dst_id,
+                url,
+                auth,
+                mode,
+                max_buffer_size: 100,
+                reconnect_delay: Duration::from_secs(1),
+                keep_alive_interval: Duration::from_secs(5),
+                keep_alive_timeout: Duration::from_secs(10),
+                connect_timeout: Duration::from_secs(20),
+                request_timeout: Duration::from_secs(20),
+                heartbeat_interval: Duration::from_secs(15),
+            },
+        }
+    }
+
+    /// Create a new client that sends messages from an Orb to an App
+    #[must_use]
+    pub fn new_as_orb(
+        url: String,
+        token: String,
+        orb_id: String,
+        session_id: String,
+    ) -> Self {
+        Self::new(
+            url,
+            Auth::Token(TokenAuth { token }),
+            orb_id,
+            session_id,
+            Mode::Orb,
+        )
+    }
+
+    /// Create a new client that sends messages from an App to an Orb
+    #[must_use]
+    pub fn new_as_app(
+        url: String,
+        token: String,
+        session_id: String,
+        orb_id: String,
+    ) -> Self {
+        Self::new(
+            url,
+            Auth::Token(TokenAuth { token }),
+            session_id,
+            orb_id,
+            Mode::App,
+        )
+    }
+
+    /// Create a new client that sends messages from an App to an Orb (using ZKP as auth method)
+    #[must_use]
+    pub fn new_as_app_zkp(
+        url: String,
+        root: String,
+        signal: String,
+        nullifier_hash: String,
+        proof: String,
+        session_id: String,
+        orb_id: String,
+    ) -> Self {
+        Self::new(
+            url,
+            Auth::ZKP(ZkpAuth {
+                root,
+                signal,
+                nullifier_hash,
+                proof,
+            }),
+            session_id,
+            orb_id,
+            Mode::App,
+        )
+    }
+
+    async fn check_for_msg<T: PayloadMatcher>(&self) -> Option<T::Output> {
+        for msg in self.get_buffered_messages().await {
+            if let Some(payload) = &msg.payload {
+                if let Some(specific_payload) = T::matches(payload) {
+                    return Some(specific_payload);
+                }
+                tracing::warn!(
+                    "While waiting for payload of type {:?}, we got: {:?}",
+                    type_name::<T>(),
+                    debug_any(&msg.payload)
+                );
+            }
+        }
+        None
+    }
+
+    /// Get buffered messages
+    pub async fn get_buffered_messages(&self) -> VecDeque<RelayPayload> {
```

Comment:
suggestion: consider having an `&mut VecDeque` as input, and using `std::mem::swap` instead of std::mem::take. This will allow the buffer to be re-used. Right now every buffer that gets built is constantly getting dropped afterwards.

If thats considered too complicated, then I would recommend buffer.clone() over std::mem::take, since this will actually be more efficient by letting the original buffer continue to grow, reducing the frequency of memory allocation.

Reply:
I got rid of VecDeque completely in the end in #342